cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaSampleStore.java

```diff
@@ -259,61 +263,82 @@ private void increaseTopicReplicationFactor(ZkUtils zkUtils,
+    if (partitionCount > topicMetadata.partitionMetadata().size()) {
+      if (!ensureTopicNotUnderPartitionReassignment(zkUtils, topic)) {
+        LOG.info("There are ongoing partition reassignments for topic {}, skip checking topic's partition count.", topic);
+        return;
+      }
+      AdminUtils.addPartitions(zkUtils, topic, partitionCount, "", true, RackAwareMode.Safe$.MODULE$);
+      LOG.info("Kafka topic " + topic + " now has " + partitionCount + " partitions.");
+    }
+  }
+
+  private void ensureTopicCreated(ZkUtils zkUtils,
+                                  Set<String> allTopics,
+                                  String topic,
+                                  long retentionMs,
+                                  int replicationFactor,
+                                  int partitionCount) {
     Properties props = new Properties();
     props.setProperty(LogConfig.RetentionMsProp(), Long.toString(retentionMs));
     props.setProperty(LogConfig.CleanupPolicyProp(), DEFAULT_CLEANUP_POLICY);
     if (!allTopics.contains(topic)) {
       AdminUtils.createTopic(zkUtils, topic, partitionCount, replicationFactor, props, RackAwareMode.Safe$.MODULE$);
     } else {
       try {
-        ensureTopicNotUnderPartitionReassignment(zkUtils, topic);
         AdminUtils.changeTopicConfig(zkUtils, topic, props);
-        MetadataResponse.TopicMetadata topicMetadata =
-            AdminUtils.fetchTopicMetadataFromZk(JavaConversions.asScalaSet(Collections.singleton(topic)), zkUtils,
-                ListenerName.forSecurityProtocol(SecurityProtocol.PLAINTEXT)).head();
-        OptionalInt currentReplicationFactor =
-            topicMetadata.partitionMetadata().stream().mapToInt(pm -> pm.replicas().size()).min();
-        if (!currentReplicationFactor.isPresent()) {
-          throw new IllegalStateException("Kafka topic " + topic + " has no partition (Metadata: " + topicMetadata + ").");
-        }
-        if (replicationFactor > currentReplicationFactor.getAsInt()) {
-          increaseTopicReplicationFactor(zkUtils, topicMetadata, replicationFactor, topic, props);
-        }
-        if (partitionCount > topicMetadata.partitionMetadata().size()) {
-          AdminUtils.addPartitions(zkUtils, topic, partitionCount, "", true, RackAwareMode.Safe$.MODULE$);
-          LOG.info("Kafka topic " + topic + " now has " + partitionCount + " partitions.");
-        }
+        MetadataResponse.TopicMetadata topicMetadata = AdminUtils.fetchTopicMetadataFromZk(JavaConversions.asScalaSet(Collections.singleton(topic)),
+                                                       zkUtils,
+                                                       ListenerName.forSecurityProtocol(SecurityProtocol.PLAINTEXT)).head();
+        maybeIncreaseTopicReplicationFactor(zkUtils, topicMetadata, replicationFactor, topic);
+        maybeIncreaseTopicPartitionCount(zkUtils, topic, topicMetadata, partitionCount);
```

Comment:
Does it matter to do RF increase after or before partition count increase?

Reply:
yes, `maybeIncreaseTopicReplicationFactor()` will write `partition_reassignment` znode,  and at the beginning of the `maybeIncreaseTopicPartitionCount()` we check for `partition_reassignment` znode  and wait for there is no ongoing partition reassignment for the same topic.